### ISSUE TITLE
Impossible to stop/cancel looping view animations

### ISSUE THREAD

**Issue by ericcorriel** [openlayers/openlayers#14421]
**Describe the bug**
```view.cancelAnimations()``` does not stop looping animations. Specifically, if ```view.animate()``` is wrapped by a function that has itself as a callback (to create a looping animation), the animation can never be stopped because ```view.cancelAnimations()``` did not account for this use case.

**To Reproduce**

1. Go to ```examples/animation.html``` and replace the html with the following to add buttons for a looping animation and ```view.cancelAnimations()```:
```
<div id="map" class="map"></div>
<button id="rotate-left" title="Rotate clockwise">↻</button>
<button id="rotate-right" title="Rotate counterclockwise">↺</button>
<button id="pan-to-london">Pan to London</button>
<button id="elastic-to-moscow">Elastic to Moscow</button>
<button id="bounce-to-istanbul">Bounce to Istanbul</button>
<button id="spin-to-rome">Spin to Rome</button>
<button id="fly-to-bern">Fly to Bern</button>
<button id="rotate-around-rome">Rotate around Rome</button>
<button id="tour">Take a tour</button>
<!-- Two new buttons here -->
<button id="never-ending-tour">Take never ending tour</button>
<button id="cancel-animation">Cancel Animation</button>
```
2. Go to ```examples/animation.js``` and add the following to create a looping animation and a call to ```view.cancelAnimations()```. Notice ```neverEndingTour``` has itself as a callback to create a looping animation.

```
function neverEndingTour() {
  const locations = [london, bern, rome, moscow, istanbul];
  const duration = 2000;
  view.animate(
      { center: london, duration: duration },
      { center: bern, duration: duration },
      { center: rome, duration: duration },
      { center: moscow, duration: duration },
      { center: istanbul, duration: duration },
      neverEndingTour // <-- callback function calling itself creates a looping animation
      );
}

// Event handlers for new buttons 
onClick('never-ending-tour', neverEndingTour);

onClick('cancel-animation', function () {
  view.cancelAnimations();
});
```
3. Load ```animation.html``` in the browser and start any animation (**Spin to Rome**, for example) and then click **Cancel Animation** and notice the animation stops 👍
4.  Click **Take Never Ending Tour**, notice the tour loops 👍
5. Click **Cancel Animation** and notice the tour doesn't stop, it only restarts from the beginning 👎
6. In ```animation.js```, replace the callback ```neverEndingTour``` with ```function(){ flyTo(bern, function () {}) }```
7.  Going back to the UI, click **Never Ending Tour** and then **Cancel Animation** and notice that the animation series is canceled and the user is immediately taken to Bern. 

In short, it seems ```view.cancelAnimations()``` was designed to cancel the animation *series* but not the animation as a whole. 

**Expected behavior**
1. In the case of a looping animation, ```view.cancelAnimations()``` should stop the animation
2. While this seems to be an intentional design decision, as demonstrated by steps (6) and (7) above, ```view.cancelAnimations()```  does not actually cancel the animation. Rather it skips the remaining animations in the series and goes directly to the callback. While (2) is not necessairly a bug, its implementation has the unfortunate side effect of creating one in the case of looping animations.

**Comment by mike-000** [comment]
You can handle cancellation in the callback as it is called with a completed argument.  If `false` the animation will have been cancelled.  Otherwise it will be `true` (or `undefined` for an initial call).

```
function neverEndingTour(completed) {
  if (completed === false) {
    return;
  }
  const locations = [london, bern, rome, moscow, istanbul];
  const duration = 2000;
  view.animate(
      { center: london, duration: duration },
      { center: bern, duration: duration },
      { center: rome, duration: duration },
      { center: moscow, duration: duration },
      { center: istanbul, duration: duration },
      neverEndingTour // <-- callback function calling itself creates a looping animation
      );
}
```

**Comment by ericcorriel** [comment]
@mike-000 nice–I did not think of that!

I'd like to propose an alternative solution: https://github.com/openlayers/openlayers/pull/14423, which I think adds conceptual clarity while also avoiding the function signature mismatch that your solution requires. Let me know what you think…

**Comment by ahocevar** [comment]
@ericcorriel Which function signature mismatch are you referring to?

**Comment by ericcorriel** [comment]
@ahocevar re: signature mismatch: I know js doesn't require parameters to always be instantiated, but in @mike-000 's example above, the callback function ```neverEndingTour``` is passed without any parameters but the function definition expects a ```completed``` parameter. Even if ```completed``` is ultimately passed by ```animationCallback()``` in ```View.js``` (line 1903), I would argue that's not exactly easy to intuit…

Pasting @mike-000 's solution, which achieves the desired functionality (stopping a looping animation) within the current framework and allows for a parameterized function wrapping the call to ```view.animate()```:

```
function neverEndingTour(firstStop) {
  const locations = [london, bern, rome, moscow, istanbul];
  const duration = 2000;
  view.animate(
    { center: firstStop, duration: duration },
    { center: bern, duration: duration },
    { center: rome, duration: duration },
    { center: moscow, duration: duration },
    { center: istanbul, duration: duration },
    function (completed) {
      if (completed) {
        neverEndingTour(
          locations[Math.floor(Math.random() * (locations.length + 1))]
        );
      }
    } 
  );
}
```

**Comment by ahocevar** [comment]
Where do you see a mismatch here? Like you said, if a callback function does not do anything with an argument provided by the caller, it is created without that argument. That's very common practice in JavaScript and even enforced by popular style guides available in linter tools.

**Comment by ericcorriel** [comment]
@ahocevar understood, I'm just expressing a personal preference against implicit arguments–definitely not a big deal!
